### PR TITLE
Feature/department

### LINKED ISSUE #3: Department - add CRUD for department feature
I want to create a feature called department.
In this feature, we need CRUD along with search.
Such documents and tests related to this feature will be added.

### PR DESCRIPTION
**Pull request by mohaphez**
All actions related to department feature are added . also all test passed .

### CODE REVIEW

File: .phpunit.result.cache [hunk 1]
```diff
@@ -0,0 +1 @@
+{"version":1,"defects":{"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testShowList":4,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testStore":4,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testSearch":4,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testShow":4,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testUpdate":4,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testDelete":4},"times":{"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testShowList":0.267,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testStore":0.043,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testSearch":0.051,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testShow":0.035,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testUpdate":0.047,"dnj\\Ticket\\Test\\Fature\\DepartmentControllerTest::testDelete":0.051}}
```
Comment (yeganemehr) [inline .phpunit.result.cache]: This file should not commit to repo. Please remove it and put it's name in .gitignore

File: composer.json [hunk 2]
```diff
@@ -1,43 +1,45 @@
...
+  "type": "library",
+  "require": {
+    "php": "^8.1"
+  },
+  "require-dev": {
+    "phpunit/phpunit": "^9.5",
+    "phpstan/phpstan": "^1.4.1",
+    "friendsofphp/php-cs-fixer": "^3.1",
+    "orchestra/testbench": "^7.0",
+    "laravel/legacy-factories": "^1.0.4"
```
Comment (yeganemehr) [inline composer.json]: What is this? can you tell me why we need this?

File: composer.json [hunk 3]
```diff
@@ -1,43 +1,45 @@
...
-    "minimum-stability": "dev"
+  "name": "dnj/laravel-ticketing",
+  "description": "Package used for the ticketing system.",
+  "type": "library",
+  "require": {
+    "php": "^8.1"
+  },
+  "require-dev": {
+    "phpunit/phpunit": "^9.5",
+    "phpstan/phpstan": "^1.4.1",
```
Comment (yeganemehr) [inline composer.json]: if you install phpstan, then you should install larastan as well. Also you should config it and make sure that it cannot report any error in max level.

File: database/factories/UserFactory.php [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+use dnj\Ticket\Models\User;
+
+$factory->define(User::class, function (Faker\Generator $faker) {
```
Comment (yeganemehr) [inline database/factories/UserFactory.php]: This package does not anything do with user.
If you need this for tests, put it in test directory.

File: docs/department.yaml [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                      id:
+                        type: integer
+                      title:
+                        type: string
+              examples:
+                Example 1:
+                  value:
+                    department:
+                      id: 12
+                      title: New Pepartment
```
Comment (yeganemehr) [inline docs/department.yaml]:
```suggestion
                      title: New Department
```

File: phpunit.xml [hunk 6]
```diff
@@ -1,36 +1,31 @@
...
-         beStrictAboutTodoAnnotatedTests="true"
-         failOnRisky="true"
-         failOnWarning="true"
-         verbose="true">
+         colors="true"
+>
     <testsuites>
-        <testsuite name="default">
-            <directory suffix="Test.php">tests</directory>
+        <testsuite name="Unit">
```
Comment (yeganemehr) [inline phpunit.xml]: There is no unit test. phpunit reports a fatal error.

File: routes/api.php [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+use dnj\Ticket\Http\Controllers\DepartmentController;
+use Illuminate\Routing\Middleware\SubstituteBindings;
```
Comment (yeganemehr) [inline routes/api.php:4]: I don't know what it is, can you explain it for me?

File: src/Http/Controllers/DepartmentController.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+use dnj\Ticket\Models\Department;
+
+use Illuminate\Routing\Controller;
+use Illuminate\Http\Request;
+
+class DepartmentController extends Controller
+{
+    public function __construct()
+    {
+        $this->middleware('auth');
```
Comment (yeganemehr) [inline src/Http/Controllers/DepartmentController.php]: please put this at routing files.

File: src/Http/Controllers/DepartmentController.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        $department->save();
+
+        return new DepartmentResource($department);
+    }
+
+    public function destroy(Department $department)
+    {
+        $department->delete();
+
+        return response([], 204);
```
Comment (yeganemehr) [inline src/Http/Controllers/DepartmentController.php]:
```suggestion
        return response()->noContent();
```

File: src/Http/Requests/DepartmentRequest.php [hunk 10]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace dnj\Ticket\Http\Requests;
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class DepartmentRequest extends FormRequest
```
Comment (yeganemehr) [inline src/Http/Requests/DepartmentRequest.php]:
```suggestion
class DepartmentUpsertRequest extends FormRequest
```

File: src/Http/Resources/DepartmentResource.php [hunk 11]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace dnj\Ticket\Http\Resources;
+
+use Illuminate\Http\Resources\Json\JsonResource;
+
+class DepartmentResource extends JsonResource
```
Comment (yeganemehr) [inline src/Http/Resources/DepartmentResource.php:7]: Why do you customize this JsonResource?
There is anything wrong with original JsonResource behavior?

**Review by yeganemehr — CHANGES_REQUESTED** [review]

File: routes/api.php (hunk 7 above)
Comment (mohaphez) [reply routes/api.php:4]: Without this middleware, the implicit route binding will not work.
In Laravel itself, it exists by default in the Http/Kernel.php file in the middleware section.

File: database/factories/UserFactory.php (hunk 4 above)
Comment (mohaphez) [reply database/factories/UserFactory.php]: I put it in the tests/factories folder

File: composer.json (hunk 2 above)
Comment (mohaphez) [reply composer.json]: When using the factory model without this package, I received an error.
Currently, I changed the factory build method in the new commit and this package is no longer needed.
I also removed it from the composer.json file

File: src/Http/Resources/DepartmentResource.php (hunk 11 above)
Comment (mohaphez) [reply src/Http/Resources/DepartmentResource.php:7]: There was no problem.
I have now reverted to the original JsonResource

**Review by yeganemehr — APPROVED** [review]

**Pull request merged**